Sources/SwiftBuildSupport/PIFBuilder.swift

```diff
@@ -20,12 +20,68 @@ import TSCUtility
 @_spi(SwiftPMInternal)
 import SPMBuildCore
 
-import func TSCBasic.memoize
 import func TSCBasic.topologicalSort
 import var TSCBasic.stdoutStream
 
 import enum SwiftBuild.ProjectModel
 
+public func memoize<T>(to cache: inout T?, build: () async throws -> T) async rethrows -> T {
+    if let value = cache {
+        return value
+    } else {
+        let value = try await build()
+        cache = value
+        return value
+    }
+}
+
+extension ModulesGraph {
+    public static func computePluginGeneratedFiles(
+        target: ResolvedModule,
+        toolsVersion: ToolsVersion,
+        additionalFileRules: [FileRuleDescription],
+        buildParameters: BuildParameters,
+        buildToolPluginInvocationResults: [PackagePIFBuilder.BuildToolPluginInvocationResult],
+        prebuildCommandResults: [CommandPluginResult],
+        observabilityScope: ObservabilityScope
+    ) throws -> (pluginDerivedSources: Sources, pluginDerivedResources: [Resource]) {
+        var pluginDerivedSources = Sources(paths: [], root: buildParameters.dataPath)
+
+        // Add any derived files that were declared for any commands from plugin invocations.
+        var pluginDerivedFiles = [AbsolutePath]()
+        for command in buildToolPluginInvocationResults.reduce([], { $0 + $1.buildCommands }) {
+            for absPath in command.outputPaths {
+                pluginDerivedFiles.append(try AbsolutePath(validating: absPath))
+            }
+        }
+
+        // Add any derived files that were discovered from output directories of prebuild commands.
+        for result in prebuildCommandResults {
+            for path in result.derivedFiles {
+                pluginDerivedFiles.append(path)
+            }
+        }
+
+        // Let `TargetSourcesBuilder` compute the treatment of plugin generated files.
+        let (derivedSources, derivedResources) = TargetSourcesBuilder.computeContents(
+            for: pluginDerivedFiles,
+            toolsVersion: toolsVersion,
+            additionalFileRules: additionalFileRules,
+            defaultLocalization: target.defaultLocalization,
+            targetName: target.name,
+            targetPath: target.underlying.path,
+            observabilityScope: observabilityScope
+        )
+        let pluginDerivedResources = derivedResources
+        derivedSources.forEach { absPath in
```

Comment:
Nit: According to the code review we did in XCBuild (before OSS), the Swift folks suggested we use plain `for` loops for such cases.